testcases/cve/stack_clash.c

```diff
@@ -0,0 +1,255 @@
+ * THRESHOLD from the mmapped area.
+ *
+ * We read /proc/self/maps to examine exact top of the stack and `mmap(2)`
+ * our region exactly GAP_PAGES * PAGE_SIZE away. We read /proc/cmdline to
+ * see if a different stack_guard_gap size is configured.
+ *
+ * [1] https://blog.qualys.com/securitylabs/2017/06/19/the-stack-clash
+ * [2] https://bugzilla.novell.com/show_bug.cgi?id=CVE-2017-1000364
+ */
+
+#include <sys/mman.h>
+#include <sys/wait.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <alloca.h>
+#include <signal.h>
+#include <stdlib.h>
+
+#include "tst_test.h"
+#include "tst_safe_stdio.h"
+
+static unsigned long PAGE_SIZE;
+static unsigned long PAGE_MASK;
+static unsigned long GAP_PAGES = 256;
+static unsigned long THRESHOLD;
+static int STACK_GROWSDOWN;
+
+#define SIGNAL_STACK_SIZE	(1UL<<20)
+#define FRAME_SIZE		1024
+
+#define EXIT_TESTBROKE		TBROK
+
+void exhaust_stack_into_sigsegv(void)
+{
+	volatile char * ptr = alloca(FRAME_SIZE - sizeof(long));
+	*ptr = '\0';
+	exhaust_stack_into_sigsegv();
+}
+
+#define MAPPED_LEN PAGE_SIZE
+static unsigned long mapped_addr;
+
+void segv_handler(int sig, siginfo_t *info, void *data LTP_ATTRIBUTE_UNUSED)
+{
+	unsigned long fault_addr = (unsigned long)info->si_addr;
+	unsigned long mmap_end = mapped_addr + MAPPED_LEN;
+	ssize_t diff;
+
+	if (sig != SIGSEGV && sig != SIGBUS)
+		return;
+
+	if (STACK_GROWSDOWN)
+		diff = fault_addr - mmap_end;
+	else
+		diff = mapped_addr - fault_addr;
+
+	tst_res(TINFO,
+		"mmap = [%lx, %lx), addr = %lx, diff = %lx, THRESHOLD = %lx",
+		mapped_addr, mmap_end, fault_addr, diff, THRESHOLD);
+	if (diff < 0 || (unsigned long)diff < THRESHOLD)
+		_exit(EXIT_FAILURE);
+	else
+		_exit(EXIT_SUCCESS);
+}
+
+unsigned long read_stack_addr_from_proc(unsigned long *stack_size)
+{
+	FILE *fh;
+	char buf[1024];
+	unsigned long stack_top = -1UL, start, end;
+
+	fh = SAFE_FOPEN("/proc/self/maps", "r");
+
+	while (!feof(fh)) {
+		if (fgets(buf, sizeof(buf), fh) == NULL) {
+			tst_brk(TBROK | TERRNO, "fgets");
+			goto out;
+		}
+
+		if (!strstr(buf, "[stack"))
+			continue;
+
+		if (sscanf(buf, "%lx-%lx", &start, &end) != 2) {
+			tst_brk(TBROK | TERRNO, "sscanf");
+			goto out;
+		}
+
+		*stack_size = end - start;
+
+		if (STACK_GROWSDOWN)
+			stack_top = start;
+		else
+			stack_top = end;
+		break;
+	}
+
+out:
+	SAFE_FCLOSE(fh);
+	return stack_top;
+}
+
+void do_child(void)
+{
+	unsigned long stack_addr, stack_size;
+	stack_t signal_stack;
+	struct sigaction segv_sig = {.sa_sigaction = segv_handler, .sa_flags = SA_ONSTACK|SA_SIGINFO};
+	void *map;
+	unsigned long gap = GAP_PAGES * PAGE_SIZE;
+	struct rlimit rlimit;
+
+	stack_addr = read_stack_addr_from_proc(&stack_size);
+	if (stack_addr == -1UL) {
+		tst_brk(TBROK, "can't read stack top from /proc/self/maps");
+		return;
+	}
+
+	SAFE_GETRLIMIT(RLIMIT_STACK, &rlimit);
+	if (rlimit.rlim_cur != RLIM_INFINITY && stack_size >= rlimit.rlim_cur) {
+		rlimit.rlim_cur = stack_size + PAGE_SIZE;
```

Comment:
From the following note on s390x it seems like it won't be enough. Or does s390x has 8KiB pages?